src/shell/shell_cmds.cpp

```diff
@@ -765,39 +960,67 @@ void DOS_Shell::CMD_LS(char *args)
 
 	do {
 		DtaResult result;
-		dta.GetResult(result.name, result.size, result.date,
-		              result.time, result.attr);
+		dta.GetResult(result.name, result.size, result.date, result.time, result.attr);
+		safe_strcpy(result.lname,result.name);
 		results.push_back(result);
 	} while ((ret = DOS_FindNext()) == true);
+	lfn_filefind_handle=fbak;
+
+	size_t w_count, p_count, col;
+	unsigned int max[10], total, tcols=real_readw(BIOSMEM_SEG,BIOSMEM_NB_COLS);
+	if (!tcols) tcols=80;
+
+	for (col=10; col>0; col--) {
+		for (int i=0; i<10; i++) max[i]=2;
+		if (optL) col=1;
+		if (col==1) break;
+		w_count=0;
+		for (const auto &entry : results) {
+			std::string name = uselfn&&!optZ?entry.lname:entry.name;
+			if (name == "." || name == "..") continue;
+			if (!optA && (entry.attr&DOS_ATTR_SYSTEM || entry.attr&DOS_ATTR_HIDDEN)) continue;
+			if (name.size()+2>max[w_count%col]) max[w_count%col]=(unsigned int)(name.size()+2);
+			++w_count;
+		}
```

Comment:
This is a dense little bundle of fun :-)  two nested loops increments, breaks, and continues, and various magic integers =1, ==1, =2, +2, and array-indexes generated from count % col.

This needs to be significantly unpacked and documented.

Reply:
@kcgen The original code (before this PR) for LS command simply fixes it to 5 columns, which won't work at all if the file names are variable especially for LFN support. The new code will do a trial and error by itself to find out the best number of columns to display the file names according the length, as in the actual Unix/Linux implementation for this command. That is why nested loops are necessary for this to work (I improved this command greatly after borrowing it from DOSBox-staging to DOSBox-X, and now I am making the changes back to DOSBox-staging). But I agree some minor improvements can be made in the code.